// @loopback/express/dist/mixins/middleware.mixin.js
"use strict";
// Copyright IBM Corp. 2020. All Rights Reserved.
// Node module: @loopback/express
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@loopback/core");
const middleware_1 = require("../middleware");
function extendsFrom(subClass, baseClass) {
    let cls = subClass;
    while (cls) {
        if (cls === baseClass)
            return true;
        cls = Object.getPrototypeOf(cls);
    }
    return false;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function MiddlewareMixin(superClass) {
    if (!extendsFrom(superClass, core_1.Context)) {
        throw new TypeError('The super class does not extend from Context');
    }
    return class extends superClass {
        /**
         * @internal
         * Implementation of `expressMiddleware`
         */
        expressMiddleware(factoryOrKey, configOrHandlers, options = {}) {
            const key = factoryOrKey;
            if (core_1.isBindingAddress(key)) {
                const handlers = Array.isArray(configOrHandlers)
                    ? configOrHandlers
                    : [configOrHandlers];
                // Create middleware that wraps all Express handlers
                if (handlers.length === 0) {
                    throw new Error('No Express middleware handler function is provided.');
                }
                return middleware_1.registerMiddleware(this, middleware_1.toMiddleware(handlers[0], ...handlers.slice(1)), {
                    ...options,
                    key,
                });
            }
            else {
                return middleware_1.registerExpressMiddleware(this, factoryOrKey, configOrHandlers, options);
            }
        }
        /**
         * Register a middleware function or provider class
         *
         * @example
         * ```ts
         * const log: Middleware = async (requestCtx, next) {
         *   // ...
         * }
         * server.middleware(log);
         * ```
         *
         * @param middleware - Middleware function or provider class
         * @param options - Middleware binding options
         */
        middleware(middleware, options = {}) {
            return middleware_1.registerMiddleware(this, middleware, options);
        }
    };
}
exports.MiddlewareMixin = MiddlewareMixin;
