// @loopback/repository/dist/types/index.js
"use strict";
// Copyright IBM Corp. 2017. All Rights Reserved.
// Node module: @loopback/repository
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Built-in types for LoopBack modeling
 * - Type: abstract base type
 * - StringType: string
 * - BooleanType: boolean
 * - NumberType: number
 * - DateType: Date
 * - BufferType: Buffer
 * - AnyType: any
 * - ArrayType: Array<T>
 * - UnionType: Union of types
 */
const any_1 = require("./any");
exports.AnyType = any_1.AnyType;
const array_1 = require("./array");
exports.ArrayType = array_1.ArrayType;
const boolean_1 = require("./boolean");
exports.BooleanType = boolean_1.BooleanType;
const buffer_1 = require("./buffer");
exports.BufferType = buffer_1.BufferType;
const date_1 = require("./date");
exports.DateType = date_1.DateType;
const model_1 = require("./model");
exports.ModelType = model_1.ModelType;
const null_1 = require("./null");
exports.NullType = null_1.NullType;
const number_1 = require("./number");
exports.NumberType = number_1.NumberType;
const object_1 = require("./object");
exports.ObjectType = object_1.ObjectType;
const string_1 = require("./string");
exports.StringType = string_1.StringType;
const union_1 = require("./union");
exports.UnionType = union_1.UnionType;
exports.STRING = new string_1.StringType();
exports.BOOLEAN = new boolean_1.BooleanType();
exports.NUMBER = new number_1.NumberType();
exports.DATE = new date_1.DateType();
exports.BUFFER = new buffer_1.BufferType();
exports.ANY = new any_1.AnyType();
exports.NULL = new null_1.NullType();
